Create tool to check for circular dependencies in source files

devversion opened the pull request:
Creates a tool for validating TypeScript circular dependencies. The tool
has been designed in a way that allows us to slowly burn down the amount
of circular dependencies while ensuring that we don't regress.

For more context. See: #35041 and [this comment](https://github.com/angular/angular/pull/34774#issuecomment-577237684).

devversion committed:
build: create tool for validating typescript circular dependencies

Creates a tool for validating TypeScript circular dependencies. The tool
has been designed in a way that allows us to slowly burn down the amount
of circular dependencies while ensuring that we don't regress.

The tool doesn't rely on Madge since it doesn't provide a programmatic
way for doing path mapping. We need path mapping since we also want to
check for cycles across different entry-points or packages. The tool
uses the TypeScript AST to manually collect cycles. This code is not
a lot of bloat and also gives us more flexibility (if we ever need it).

Closes #35041.

devversion committed:
ci: set up golden file for typescript circular dependencies

Sets up a golden file for the TypeScript circular dependencies for
source files inside of the `packages/` folder.

Also sets up the appropriate Yarn shorthand scripts, and a codeowner
group that is soley responsible for verifying changes to the golden.

devversion committed:
ci: run ts-circular-deps check in lint job on ci

josephperrott commented on tools/ts-circular-dependencies/index.ts:
```diff
@@ -0,0 +1,156 @@
+  if (isMatching) {
+    console.info(chalk.green('✅  Golden matches current circular dependencies.'));
+    console.info(
+        chalk.green(`   Current amount of cycles: ${chalk.yellow(cycles.length.toString())}`));
```
Can we change this to be "Current number of cycles".

Additionally, can we always print out the number of cycles in the repo whenever the tool is run.

josephperrott reviewed (commented):
Before we approve and commit this, we are also talking about if this should actually live in dev-infra repo and then be depended on via npm.  So we can do the same process for components and tooling.

devversion commented:
@josephperrott I heard about these plans in the components sync. Concerns about not having another repository for this stuff had been raised. Also personally I'm slightly concerned about going through NPM since it could end up as with `ts-api-guardian` where it's difficult/non-standard process to get a release. It might be good to just use these things through a Bazel `http_repository`.

There are surely benefits/downsides but the primary benefit would be that we don't need any `builds` repositories for `dev-infra` and that we could easily use shared Bazel rules (like `npm_integration`).

Anyway though; TL;DR: Is there anything I need to do? Should I move this already over, or should we just get this started now and move it over once we decided on the process?

devversion commented on packages/circular-dependencies-golden.json:
```diff
@@ -0,0 +1,5285 @@
+[
```
Done. Added a small readme. I think the convention should be to have a description for each golden we add there. WDYT?

devversion committed:
refactor: move ts-circular-dependencies tool to dev-infra

devversion committed:
build: remove rollup packaging from dev-infra

The dev-infra package currently uses rollup for packaging. This has been
done initially as a way to workaround manifest paths being used in the
AMD JavaScript output.

The actual solution to this problem is setting module names that match
the `package.json` name. This ensures that the package can be consumed
correctly in Bazel, and through NPM. This allows us to get rid of the
rollup bundling, and we don't need to hard-code which dependencies
should be external or included.

Additionally, tools that are part of `dev-infra` can now specify
their external dependencies simply in the `package.json`. To reduce
version duplication, and out-of-sync versions, a new genrule has been
created that syncs the versions with the top-level project
`package.json`.

devversion commented on package.json:
```diff
@@ -34,7 +34,10 @@
     "public-api:check": "node goldens/public-api/manage.js test",
-    "public-api:update": "node goldens/public-api/manage.js accept"
+    "public-api:update": "node goldens/public-api/manage.js accept",
+    "ts-circular-deps": "ts-node dev-infra/ts-circular-dependencies/index.ts",
```
@josephperrott For now, until your packaging PR lands, I think we should be also fine just using `ts-node` here. It's minimal and allows us to test the changes immediately. We can always change that single line when we have the long-term plan.

devversion replied:
The generic one is actually used so that the `ts-node <..>` command doesn't need to be repeated all the time.

josephperrott commented on dev-infra/BUILD.bazel:
```diff
@@ -1,38 +1,38 @@
     visibility = ["//visibility:public"],
     deps = [
-        ":bundle",
+        ":cli",
```
Moving to a dependency on the ts_library outputs results in requiring the full umd name which doesn't work as needed.

For instance if you build the `npm_package` current and add it to try to run:
```sh
$ yarn -s ng-dev
/Users/jperrott/js/angular/node_modules/.bin/ng-dev: line 1: syntax error near unexpected token `('
/Users/jperrott/js/angular/node_modules/.bin/ng-dev: line 1: `(function (factory) {'
```

We might have to wait on https://github.com/bazelbuild/rules_typescript/pull/492 for us to be able to do something like this.

devversion replied:
I've explained in the commit message what the original problem was. There is nothing wrong with having the right AMD module paths. That's actually how the Angular NPM package output works too. The AMD module names just need to match the module name of the actual NPM package.

Also I added the shebang, so that it can be launched without having to use `node <script>`.

josephperrott reviewed (approved):
LGTM
